---
title: "mkchrannot"
output: html_document
---

```{r setup, echo=FALSE, warning=FALSE}

library(tidyverse)
library(dplyr)
library(slider)
library(ggplot2)
library(readxl)
library(liftOver)
library(annotatr)
library(GenomicRanges)
library(GenomicFeatures)
library(data.table)
library(xlsx)
library(rtracklayer)
require(readr)
library(ggridges)
library(genomation)

```

```{r func, echo=FALSE, warning=FALSE}
read_pwd <- function(x){
  dat <- fread(x, header = F)
  colnames(dat) <- c("chr", "start", "end", "pwd", "beta.x", "beta.y")
  #dat <- dat %>% filter(chr=="chr19")
  dat <- as.data.frame(dat)
  return(dat)
}

# Annotation function
annot <- function(dat, region){
  wgpos <- GRanges(dat)
  dm_annotated <- annotate_regions(
    regions = wgpos,
    annotations = region,
    ignore.strand = TRUE,
    quiet = FALSE)
  
  dat_annot <- data.frame(dm_annotated)
  
  return(dat_annot)
}

# make chromosome region -- 100bp bins
mkchrreg <- function(howlong, whichchr){
  a <- rep(0, howlong)
  b <- rep(0, howlong)
  i <- 1
for(i in 1:howlong){
  a[i] <- (i-1)*100+1
  b[i] <- i*100
  i+1
}

table1 <- cbind(a, b)
colnames(table1) <- c("start", "end")

chrreg <- as.data.frame(table1) %>% mutate(seqnames=whichchr, regid=row_number()) %>% 
  dplyr::select(seqnames, start, end, regid)

return(chrreg)
}

# annot gene region
annotregion <- function(whichchr){
  
annot_chr <- annot(whichchr, gene_region) %>% 
  distinct(seqnames, start, end, regid, annot.gene_name, annot.transcript_id, annot.type)

# locate intron
annotchr_sub <- annot_chr %>% filter(annot.type=="transcript") %>% 
  distinct(seqnames, start, end, regid, annot.gene_name, annot.transcript_id) 

annotintron <- annot(annotchr_sub, all.introns) %>% 
  dplyr::select(seqnames, start, end, regid, annot.gene_name, annot.transcript_id, annot.tx_name, annot.gene_id) %>% 
   mutate(flag=0)

for(i in 1:length(annotintron$start)){
  annotintron$flag[i]=ifelse(
    grepl(annotintron$annot.gene_name[i], annotintron$annot.gene_id[i], fixed = TRUE)
   &grepl(annotintron$annot.transcript_id[i], annotintron$annot.tx_name[i], fixed = TRUE), 1, 0)
}

annot_intron <- annotintron %>% filter(flag==1) %>% 
  mutate(annot.type="intron") %>% 
   distinct(seqnames, start, end, regid, annot.gene_name, annot.transcript_id,  annot.type)

annotchr_generegion <- rbind(annot_chr, annot_intron) %>% filter(annot.type!="transcript") %>% 
  distinct(seqnames, start, end, regid, annot.gene_name, annot.type)

return(annotchr_generegion)
}

```


```{r importdata, echo=FALSE, warning=FALSE}
wgdf14pwd <- read_pwd("~/bismarkData/bed/wgdf14cpwdm.bed")
wgdf23pwd <- read_pwd("~/bismarkData/bed/wgdf23cpwdm.bed")
wgdf56pwd <- read_pwd("~/bismarkData/bed/wgdf56cpwdm.bed")
wgdf12pwd <- read_pwd("~/bismarkData/bed/wgdf12pwdm.bed")
wgdf34pwd <- read_pwd("~/bismarkData/bed/wgdf34pwdm.bed") 
wgdf56npwd <- read_pwd("~/bismarkData/bed/wgdf56pwdm.bed") 
ENJNpwd <- read_pwd("~/bismarkData/bed/ENJNpwd.bed")
ENINpwd <- read_pwd("~/bismarkData/bed/ENINpwd.bed")
EAEBpwd <- read_pwd("~/bismarkData/bed/EABpwd.bed")
FAFBpwd <- read_pwd("~/bismarkData/bed/FABpwd.bed")
KAKBpwd <- read_pwd("~/bismarkData/bed/KABpwd.bed")
PAPBpwd <- read_pwd("~/bismarkData/bed/PABpwd.bed")
SASBpwd <- read_pwd("~/bismarkData/bed/SABpwd.bed")
XAXBpwd <- read_pwd("~/bismarkData/bed/XABpwd.bed")



generef <- readGFF("../DepMap/hg38.ncbiRefSeq.gtf.gz")

uniquetranscript <- generef %>% filter(type=="transcript") %>% 
  mutate(region_length=ifelse(end==start, 2, end-start+1)) %>% 
   group_by(gene_id) %>% 
    mutate(maxtran = max(region_length)) %>% 
     filter(region_length==maxtran) %>% ungroup %>% 
       group_by(gene_id) %>% mutate(rank=row_number()) %>% filter(rank==1) %>% 
        dplyr::select(gene_id, transcript_id)

generef2 <- generef %>% right_join(uniquetranscript,by=c("gene_id", "transcript_id")) %>% mutate(region_length=ifelse(end==start, 2, end-start+1)) %>% 
  dplyr::select(seqid, start, end, region_length, source, type, everything())

gene5UTR <- generef2 %>% dplyr::filter(type=="5UTR")

geneTSS200 <- gene5UTR %>% mutate(tss_start = ifelse(strand=="+", start - 200, end+1),
                                  tss_end = ifelse(strand=="+", start -1, end +200),
                                  region_length = tss_end - tss_start +1, 
                                  tss_type = "TSS200") %>% 
               dplyr::select(-start, -end, -type) %>% 
                mutate(start=tss_start, end=tss_end, type=tss_type) %>% 
                 dplyr::select(-tss_start, -tss_end, -tss_type) %>%
                  dplyr::select(seqid, start, end, region_length, source, type, everything())

geneTSS2000 <- gene5UTR %>% mutate(tss_start = ifelse(strand=="+", start - 2000, end + 201),
                                  tss_end = ifelse(strand=="+", start - 201, end + 2000),
                                  region_length = tss_end - tss_start +1, 
                                  tss_type = "TSS2000") %>% 
               dplyr::select(-start, -end, -type) %>% 
                mutate(start=tss_start, end=tss_end, type=tss_type) %>% 
                 dplyr::select(-tss_start, -tss_end, -tss_type) %>%
                  dplyr::select(seqid, start, end, region_length, source, type, everything())

generef3 <- rbind(generef2, geneTSS200, geneTSS2000) %>% filter(type!="CDS")
table(generef3$type)
gene_region <- GRanges(generef3)

txdb <- makeTxDbFromGFF("../DepMap/hg38.ncbiRefSeq.gtf.gz")
## get intron information
all.introns <- intronicParts(txdb)

```

```{r, mkchr2regions, echo=FALSE, warning=FALSE}

howlonglist <- list(2500000, 2422000, 1982000, 1903000, 1815000, 1708000, 1594000, 1451000, 1384000, 1338000,
                    1351000, 1333000, 1144000, 1069000, 1020000, 902300, 832500, 802700, 590000, 643400, 467000, 508100)

chrreglist <- list()

for (i in 1:22){
  chrreglist[[i]] <- mkchrreg(howlonglist[[i]], sprintf("chr%s", i))
  
}

# chrreg <- bind_rows(chrreglist)
# head(chrreg)
# table(chrreg$seqnames)

```



```{r mergedata1, echo=FALSE, warning=FALSE}

# merge chr region, pwd and gene annotation

comparelist <- list(wgdf14pwd, wgdf23pwd, wgdf56pwd, wgdf12pwd, wgdf34pwd, wgdf56npwd,ENJNpwd,ENINpwd, EAEBpwd,FAFBpwd, KAKBpwd,PAPBpwd,SASBpwd,XAXBpwd)
namelist <- list("N1N2", "JAJB", "IAIB", "DADB", "HAHB", "MAMB", "ENJN", "ENIN", "EAEB", "FAFB", "KAKB", "PAPB", "SASB", "XAXB")



#olddatalist <- list()
#for (i in 1:22){
#  olddatalist[[i]] <- read.csv(sprintf("../chrdata/ChrCSV/Chr%s_data.csv", i), header = TRUE, stringsAsFactors=FALSE) %>% 
#  dplyr::select(seqnames, start, end, regid, annot.type, annot.gene_name)
#}

pwdlist=list()
chrannot=list()

for (i in 1:22){
  print(i)
# merge pwd
for(k in 1:length(namelist)){
  print(k)
  chrregion <- GRanges(chrreglist[[i]])

  wktmp <- annot(comparelist[[k]], chrregion) %>% group_by(annot.regid) %>% 
  mutate(n_sites=length(start),
          ave_pwd=mean(pwd, na.rm = TRUE),
           ave_beta1 = mean(beta.x, na.rm = TRUE),
            ave_beta2 = mean(beta.y, na.rm = TRUE)) %>% ungroup %>% 
  distinct(annot.seqnames, annot.start, annot.end, annot.regid, n_sites, ave_pwd, ave_beta1, ave_beta2) 
  
  colnames(wktmp) <- c("seqnames", "start", "end", "regid", sprintf("n_sites_%s", namelist[[k]]), sprintf("ave_pwd_%s", namelist[[k]]),sprintf("ave_beta1_%s", namelist[[k]]), sprintf("ave_beta2_%s", namelist[[k]]))
   
   pwdlist[[k]] <- chrreglist[[i]] %>% left_join(wktmp, by=c("seqnames", "start", "end", "regid"))
}

mergepwd <- bind_cols(pwdlist) %>% 
  dplyr::select(seqnames...1, start...2, end...3, regid...4, starts_with("n_sites_"), starts_with("ave_pwd_"), starts_with("ave_beta1_"), starts_with("ave_beta2_")) 

colnames(mergepwd)[1:4] <- c("seqnames", "start", "end", "regid")
  
# final merge
chrannot[[i]] <- chrreglist[[i]] %>% 
  left_join(mergepwd, by=c("seqnames", "start", "end", "regid")) #%>% 
   #left_join(olddatalist[[i]], by=c("seqnames", "start", "end", "regid"))

}


#annotate the chromosomes
chrannot_n <- lapply(chrannot,annotregion)


#final merge annot with data
for (i in 1:length(chrannot_n)){
  chrannot_n[[i]] <- chrannot[[i]] %>%left_join(chrannot_n[[i]],by=c("seqnames", "start", "end", "regid"))
}


ck <- chrannot_n[[1]]
summary(ck)

saveRDS(chrannot_n, file = "../chrdata/chrannot.Rds")

```

















